frontend/src/citizen-frontend/calendar/HolidayModal.tsx

```diff
@@ -46,26 +141,72 @@ export const HolidayModal = React.memo(function HolidayModal({
         holidayPeriods,
         ([holidayPeriod]) =>
           holidayPeriod && (
-            <>
-              <div>
+            <HolidayContainer>
+              <HolidaySection>
                 <div>{holidayPeriod.description[lang]}</div>
                 <ExternalLink
                   text={i18n.calendar.holidayModal.additionalInformation}
                   href={holidayPeriod.descriptionLink[lang]}
                   newTab
                 />
-              </div>
-              {availableChildren
-                .map(
-                  (child) =>
-                    child.preferredName || child.firstName.split(' ')[0]
+              </HolidaySection>
+              {Object.keys(form.selections).map((childId) => {
+                const child = childWithId(childId)
+                return (
+                  child && (
+                    <HolidaySection key={childId}>
+                      <HolidaySelector
+                        key={childId}
+                        child={child}
+                        freeAbsencePeriod={holidayPeriod?.freePeriod}
+                        value={form.selections[childId].selectedFreePeriod}
+                        onSelectPeriod={(dateRange) =>
+                          selectFreePeriod(childId, dateRange)
+                        }
+                      />
+                    </HolidaySection>
+                  )
                 )
-                .join(', ')}
-            </>
+              })}
+            </HolidayContainer>
           )
       )}
     </AsyncFormModal>
   )
 })
 
+const HolidayContainer = styled(FixedSpaceColumn)`
+  gap: ${defaultMargins.s};
+`
+const HolidaySection = styled.div`
+  background: white;
+`
+
+type SelectedHolidays = {
+  selectedFreePeriod: FiniteDateRange | null
+}
+type HolidayForm = {
+  selections: Record<string, SelectedHolidays>
+}
+
+const initializeForm = (children: ReservationChild[]): HolidayForm => ({
+  selections: children.reduce(
+    (map, child) => ({ ...map, [child.id]: { selectedFreePeriod: null } }),
+    {}
+  )
+})
+
+const postHolidays = (form: HolidayForm): Promise<Array<Result<void>>> => {
+  const requests: AbsenceRequest[] = Object.keys(form.selections)
```

Comment:
Using `Object.entries` instead of `keys` would simplify the logic as and remove the key lookups.